Add support for booting FEL mode on sunxi platforms

trini opened the pull request:
- The sunxi-fel tool from the sunxi-tools repository must be installed.
- We only support 1 board in FEL mode at a time due to limitations of
  the sunxi-fel tool.
- The digital-loggers reset script will now check for a recovery
  implementation and if found, use it.

Tested on my A20-OLinuXino-Lime2.

Signed-off-by: Tom Rini <trini@konsulko.com>

swarren commented on bin/recovery.sunxi-fel:
```diff
@@ -0,0 +1,26 @@
+# one board will be in FEL mode at a given time (or baring that, the first
+# available board is the one we are to test
+
+sunxi-fel uboot "${U_BOOT_BUILD_DIR}"/u-boot-sunxi-with-spl.bin
```
I assume this downloads the U-Boot binary over USB/... If so, this file should be name download.sunxi-fel not recovery.sunxi-fel.

The naming convention is:

recovery.xxx gets the board into recovery mode; a mode in which you can download a binary to the board.
download.xxx communicates with a board that's in recovery mode, and actually sends the binary.

In reset.tegra-usb-recovery, you'll see a call to recovery.\* which calls a utility to manipulate reset and recovery-mode buttons/signals, followed by a call to download.\* to send the binary using e.g. tegra-uboot-flasher.

trini replied:
On Fri, Aug 26, 2016 at 01:25:58PM -0700, Stephen Warren wrote:

> > +# all copies or substantial portions of the Software.
> > +#
> > +# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
> > +# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
> > +# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
> > +# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
> > +# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
> > +# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
> > +# DEALINGS IN THE SOFTWARE.
> > +
> > +# The sunxi-fel tool supports passing a bus:dev to use.  However this value
> > +# is not stable and may change over resets.  So at this time we assume that only
> > +# one board will be in FEL mode at a given time (or baring that, the first
> > +# available board is the one we are to test
> > +
> > +sunxi-fel uboot "${U_BOOT_BUILD_DIR}"/u-boot-sunxi-with-spl.bin
> 
> I assume this downloads the U-Boot binary over USB/... If so, this
> file should be name download.sunxi-fel not recovery.sunxi-fel.

Yeah, it is.  OK, renamed.

swarren commented on bin/reset.digital-loggers:
```diff
@@ -20,3 +20,5 @@
 curl --data ${power_port}=CCL -o /dev/null --silent \
 	http://${power_user:-admin}:${power_pass:-1234}@${power_ip}/outlet
+
+[ ! -z "${recovery_impl}" ] && . "${bin_dir}/recovery.${recovery_impl}" || true
```
While I'm happy to see the ?: operator in expressions, I'd far rather see a full if []; then ...; fi for code-flow.

trini replied:
On Fri, Aug 26, 2016 at 01:26:27PM -0700, Stephen Warren wrote:

> > @@ -20,3 +20,5 @@
> > 
> >  curl --data ${power_port}=CCL -o /dev/null --silent \
> >     http://${power_user:-admin}:${power_pass:-1234}@${power_ip}/outlet
> > +
> > +[ ! -z "${recovery_impl}" ] && . "${bin_dir}/recovery.${recovery_impl}" || true
> 
> While I'm happy to see the ?: operator in expressions, I'd far rather see a full if []; then ...; fi for code-flow.

OK, fixed.

trini commented:
OK, I think with just a regular old git push -f I've updated this PR with a new version that addresses your comments (and also for the flashair one).

swarren commented on bin/download.sunxi-fel:
```diff
@@ -0,0 +1,26 @@
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+# The sunxi-fel tool supports passing a bus:dev to use.  However this value
```
It is possible to grovel in sysfs and convert a "USB bus path" (ctrl-port.port.port) into the current USB device ID. I _was_ going to write such a conversion script and use it various places in this repo, but ended up adding direct support into the tools I was using for USB bus path directly and/or using udev scripts to create a constantly-named /dev symlink instead. For reference, see:

https://github.com/NVIDIA/tegrarcm/commit/cfb923c10000c5ff9925c65706992aae56422bfe
https://github.com/swarren/uboot-test-hooks/blob/master/udev/swarren-lx1/99-ubtest.rules

Anyway, this is all fine for now; the above is an FYI.

trini replied:
Yeah, I'm not thrilled with how this part works out either, as I look at the prospect of adding in PineA64 to my setup, so I would have 2 FEL boards, possibly with both having a power leak (haven't confirmed PineA64 yet).  What I've done so far, and do use for DFU testing is https://patchwork.ozlabs.org/patch/663696/ which is just a copy of the serial by-path udev rules (it's a little buggy as I see it's picked up some other devices) since at least with DFU I can pass in that directly.  I have the notion currently to use the above, add in some comments and shell my way back to a bus:dev to use.

swarren commented on bin/u-boot-test-console:
```diff
@@ -32,4 +32,9 @@ if [ -z "${console_baud}" ]; then
 fi
 
+# Force a reset into FEL mode
+if [ "${recovery_impl}" == "sunxi-fel" ]; then
```
Hmm. Not sure what I think about this. It's adding type-specific logic into a generic file. Also, since this runs picocom after the echo, it risks (a) not capturing any output generated from processing the reset command (b) not having configured the serial device for the correct baud rate when sending "reset".

BTW, what is the echo command communicating with; a console mux daemon, the Sunxi device's boot ROM, ...?

trini replied:
So, I've dropped this part locally since it's not reliable enough.  The problem here isn't sunxi specific, but rather board design specific and commonly seen on sunxi designs (I encountered this first elsewhere even).  The problem is that due to layout issues enough power can leak over the UART to keep the CPU on even when main power is turned off.  It's not enough power for the system to be at all stable but it is enough to keep it from going off.  So the above is a hack to tell U-Boot to reset the board which in turn drops is into the ROM being in FEL mode and thus when we power the board up again for the next tests it's in the correct state.

I've replaced this locally by just saying that my "ensure everything got powered off" script also tells U-Boot on this board to reset.

swarren replied:
I still see this FEL-specific logic in bin/u-boot-test-console, both in the latest version of this pull request, and in branch add-sunxi-fel in your repo.

trini commented:
OK, I'll close this one out and re-try things at some point.  Thanks!
